pkg/authentication/token/jwt/jwt.go

```diff
@@ -0,0 +1,159 @@
+/*
+ * Tencent is pleased to support the open source community by making TKEStack
+ * available.
+ *
+ * Copyright (C) 2012-2023 Tencent. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package jwt
+
+import (
+	"context"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+	"k8s.io/klog/v2"
+
+	"tkestack.io/kstone/pkg/authentication"
+	"tkestack.io/kstone/pkg/authentication/authenticator/bearertoken"
+	"tkestack.io/kstone/pkg/controllers/util"
+)
+
+const (
+	ProviderName = "jwt"
+)
+
+var (
+	// defaultTTL will be used when a 'ttl' is not specified
+	defaultTTL = 24 * time.Hour
+	once       sync.Once
+	instance   *TokenAuthenticator
+)
+
+type TokenAuthenticator struct {
+	signMethod jwt.SigningMethod
+	key        interface{}
+	ttl        time.Duration
+}
+
+type authInfo struct {
+	username string
+	password string
+}
+
+func init() {
+	authentication.RegisterTokenFactory(ProviderName,
+		func(ctx *authentication.TokenContext) (authentication.Token, error) {
+			return initTokenInstance(ctx)
+		},
+	)
+}
+
+func initTokenInstance(ctx *authentication.TokenContext) (*TokenAuthenticator, error) {
+	var err error
+	once.Do(func() {
+		var duration time.Duration
+		var key *rsa.PrivateKey
+		if ctx.TTL == "" {
+			duration = defaultTTL
+		} else {
+			duration, err = time.ParseDuration(ctx.TTL)
+		}
+
+		key, err = jwt.ParseRSAPrivateKeyFromPEM([]byte(ctx.PrivateKey))
+		instance = &TokenAuthenticator{
+			signMethod: jwt.GetSigningMethod(ctx.SignMethod),
+			ttl:        duration,
+			key:        key,
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return instance, nil
+}
+
+func (a *TokenAuthenticator) AuthenticateToken(ctx context.Context, token string) (*authentication.Response, bool, error) {
+	info, err := a.info(token)
+	if err != nil {
+		return authentication.UnauthenticatedResponse(), false, err
+	}
+
+	authenticator, err := authentication.GetAuthenticatorProvider(bearertoken.ProviderName, &authentication.AuthenticatorContext{KubeCli: util.NewSimpleClientBuilder("").ClientOrDie()})
```

Comment:
The implementation here is too complicated, and there is a loop.
request -> generic authenticator.Request -> beartoken authenticator.Request ->  jwt.AuthenticateToken ->  beartoken authenticator.Request